### PR TITLE
[Messenger] Allow interfaces to be type-hinted as well

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27076
| License       | MIT
| Doc PR        | ø

Interfaces can be type-hinted as well for the message handlers.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Handler/Locator/ContainerHandlerLocator.php [hunk 1]
```diff
@@ -30,12 +30,27 @@ public function __construct(ContainerInterface $container)
...
+
+    private function resolveFromClass($class): ?callable
+    {
+        $handlerKey = 'handler.'.$class;
+        if ($this->container->has($handlerKey)) {
+            return $this->container->get($handlerKey);
+        }
+
+        foreach (class_implements($class, false) as $interface) {
+            if ($handler = $this->resolveFromClass($interface)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Handler/Locator/ContainerHandlerLocator.php]: not sure this needs to be recursive: `class_implements` already returns all implemented interfaces
also, what about parents?

Comment (sroze) [reply src/Symfony/Component/Messenger/Handler/Locator/ContainerHandlerLocator.php]: No point of having it recursive, that's true; removed it and added handling of its parents (and the relevant tests, obsly 💃)

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sroze.

**Comment by nicolas-grekas** [comment]
This BR is breaking the C.I. for the 4.1 branch, see e.g. https://travis-ci.org/symfony/symfony/jobs/421969557

**Pull request merged**